storage/innobase/handler/ha_innodb.cc

```diff
@@ -13564,8 +13564,9 @@ int ha_innobase::delete_table(const char *name)
     {
       dict_sys.freeze(SRW_LOCK_CALL);
       for (const dict_foreign_t* f : table->referenced_set)
-        if (dict_table_t* child= f->foreign_table)
-          if ((err= lock_table_for_trx(child, trx, LOCK_X)) != DB_SUCCESS)
+        if (dict_table_t *child= f->foreign_table)
+          if ((err= lock_table_for_trx(child, trx, LOCK_X,
+                                       LOCK_TABLE_WAIT_DICT)) != DB_SUCCESS)
             break;
       dict_sys.unfreeze();
```

Comment:
The shared `dict_sys.latch` is needed here, because concurrent `erase()` operations on `dict_table_t::referenced_set` are covered by exclusive `dict_sys.latch`. With this patch, we are temporarily releasing `dict_sys.latch` inside `lock_table_for_trx()`. According to https://cplusplus.com/reference/set/set/erase/ this would make the `for` loop unsafe:
>The elements removed are modified. Concurrently accessing other elements is safe, although iterating ranges in the container is not.

I think that these loops have to be rewritten. We must acquire references to the tables and copy them to another container, then release the `dict_sys.latch` and then wait for the table locks. In that way, there is no need to modify `lock_table_for_trx()` either.